### DISCUSSION TITLE
Allow to set a default locale using environment variables

### DISCUSSION THREAD

**Discussion by luanmm** [directus/directus#7062] [Feature Requests]
Hello, people!

I'm here again, now to open a discussion about the capability of offering a default locale configuration. I couldn't find anything like this already implemented, so that's why I'm bringing this as an idea.

What I have in mind, at least for now, is that the system could have a new environment variable to allow users to set the default locale (or maybe there is a better standardized way to get from OS, but I don't think there is something so reliable that let us work with).

In my context, I'm using Directus in other language than "en-US" (which is the default one). But when I set up the custom project (which will be deployed from a repository), after the first login and every new user that I create has the need to be changed to the desired language. With a default locale, the system could adapt some things in a customized way (e.g. new users would be given a default language based on this configuration).

I may submit a PR after some discussion and approval from the team.

**Comment by benhaynes** [comment]
Seems like this could be a useful feature. I defer to @rijkvanzanten on if and how this should be implemented.

**Comment by rijkvanzanten** [comment]
Makes sense to me! 

The biggest question becomes: What language should the login page be in. Ideally, that's pulled from the browser 🤔

**Reply by luanmm** [reply]
Agreed that the prioritized locale should be pulled from the client/browser (maybe even to set the default language for new users, when created via browser). But this proposed configuration would be used exclusively by the server to handle cases where locale should be known and the user is not involved directly (mostly for internal setup or maybe in other processed events).

In my scenario, the only case that offers a "non-ideal" experience for now (talking about the configuration/setup) is when I am able to set everything up in the customized project repository (I have some custom endpoints, migrations for database structure and a simple mailing service with a custom template), but there is no way to set the "default language" for users, so the first admin user is created with default language (english).

This is not a good experience in my project because for a specific language I have customizations, but not for english (because of the target audience, that does not demand it). So this is the only point in the system that could not be "ready" just bootstrapping Directus (without manual configuration to change admin language).

Searching a little bit, I've found that maybe we could get the default locale from the server (just like the browser's locale is being "automatically" got for Date objects and so on) using `Intl.DateTimeFormat().resolvedOptions().locale`. This way Directus respects system (and user, when in browser context), getting locale from the "right" source in both sides.

If this runtime option could not be resolved (Node can be [built without internationalization support](https://nodejs.org/api/intl.html#intl_options_for_building_node_js)), the default could be "en-US", like today.

Do you see other uses for locale besides user/admin creation, @rijkvanzanten? Or this scenario is the only one inside Directus where a locale should be used? This question is specific to allow me to think about what should be changed if this discussion leads us to a PR.

**Reply by rijkvanzanten** [reply]
Makes sense! 

In any case

> If this runtime option could not be resolved (Node can be built without internationalization support)

We shouldn't rely on Node's current "datetime", as that will most likely be set to a default that'll probably be en-US and not really match the users wanted locale. 

> Do you see other uses for locale besides user/admin creation, @rijkvanzanten?

It's default language for newly created users (like you mentioned), and the language used on the login/reset-password pages for now. In the future, it's very likely it would also be used for translated emails like user-invite and password-reset

**Reply by luanmm** [reply]
> We shouldn't rely on Node's current "datetime", as that will most likely be set to a default that'll probably be en-US and not really match the users wanted locale.

Do you think that an environment variable would be a better option, then?

Thanks for the clarification and opinions! :fist_oncoming:

**Reply by rijkvanzanten** [reply]
We could do an env var (or maybe a setting in directus_settings) for the default language for new users, and then look into pulling the current browser locale for use on the public pages??

**Reply by nickrum** [reply]
How about adding a "project language" setting to `directus_settings` and falling back to that if the users language is set to `null`? That way the default language could be changed after the roles have been created.

**Reply by christianrr** [reply]
I've started a pull request to add the discussed functionality (#8196). Seems to work as intended ✌️

Beyond that, I also implemented the use of the project_language as default language for new users (by adapting the users module). I am not 100% sure if this is the best way to do it, but at least it works...

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
